internal/extender/failover.go

```diff
@@ -0,0 +1,296 @@
+	werror "github.com/palantir/witchcraft-go-error"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	corelisters "k8s.io/client-go/listers/core/v1"
+)
+
+// SyncResourceReservationsAndDemands gets all resource reservations and pods,
+// finds all pods that do not have a claimed spot in their resource reservation object,
+// groups them by spark application id, and patched the resource reservation objects so
+// they now reflect the current state of the world. This is needed on a leader failover,
+// as async writes for resource reservation objects mean some writes will be lost on
+// leader change, so the extender needs to call this before accepting requests.
+func SyncResourceReservationsAndDemands(
+	ctx context.Context,
+	podLister corelisters.PodLister,
+	nodeLister corelisters.NodeLister,
+	resourceReservations *cache.ResourceReservationCache,
+	demands *cache.SafeDemandCache,
+	overheadComputer *OverheadComputer) error {
+	pods, err := podLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	nodes, err := nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	rrs := resourceReservations.List()
+	availableResources, orderedNodes := availableResourcesPerInstanceGroup(ctx, rrs, nodes, overheadComputer.GetOverhead(ctx, nodes))
+	staleSparkPods := sparkPodsByAppID(ctx, rrs, pods)
+
+	r := &reconciliator{podLister, resourceReservations, demands, availableResources, orderedNodes}
+	for _, sp := range staleSparkPods {
+		r.syncResourceReservation(ctx, sp)
+		r.syncDemand(ctx, sp)
+	}
+	// recompute overhead to account for newly created resource reservations
+	overheadComputer.compute(ctx)
+	return nil
+}
+
+// sparkPods is a collection of stale state, it is comprised of
+// pods from a spark application that do not have a claimed resource reservation,
+// and the last known state of the resource reservation object
+type sparkPods struct {
+	driver    *v1.Pod
+	executors []*v1.Pod
+}
+
+type reconciliator struct {
+	podLister            corelisters.PodLister
+	resourceReservations *cache.ResourceReservationCache
+	demands              *cache.SafeDemandCache
+	availableResources   map[string]resources.NodeGroupResources
+	orderedNodes         map[string][]*v1.Node
+}
+
+func (r *reconciliator) syncResourceReservation(ctx context.Context, sp *sparkPods) {
+	// if the driver is nil it already has an associated reservation, get the resource
+	// reservation object and update it so it has reservations for each stale executor
+	if sp.driver == nil && len(sp.executors) > 0 {
+		exec := sp.executors[0]
+		rr, ok := r.resourceReservations.Get(exec.Namespace, exec.Labels[SparkAppIDLabel])
+		if !ok {
+			logRR(ctx, "resource reservation deleted, ignoring", exec.Namespace, exec.Labels[SparkAppIDLabel])
+			return
+		}
+		err := r.patchResourceReservation(sp.executors, rr.DeepCopy())
```

Comment:
not clear to me why you need a Deep Copy

Reply:
you can't change objects you got from the informer / cache